components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/JDBCScopeValidator.java

```diff
@@ -224,22 +239,117 @@ private boolean validateScope(String[] requestedScopes, AuthenticatedUser user)
             return true;
         }
 
+        String[] userRoles = null;
         int tenantId = getTenantId(user);
-        String[] userRoles = getUserRoles(user);
 
-        for (String scope : requestedScopes) {
-            if (!isScopeValid(scope, tenantId)) {
-                //If the scope is not registered return false
-                log.error("Requested scope " + scope + " is invalid");
-                return false;
+        /*
+        Here we handle scope validation for federated user and local user separately.
+        For local users - user store is used to get user roles.
+        For federated user - get user roles from user attributes.
+        Note that if there is association between a federated user and local user () 'Assert identity using mapped local
+        subject identifier' flag will be set as true. So authenticated user will be associated local user not
+        federated user.
+         */
+        if (user.isFederatedUser()) {
+            /*
+            There is a flow where 'Assert identity using mapped local subject identifier' flag enabled but the
+            federated user doesn't have any association in localIDP, to handle this case we check for 'Assert
+            identity using mapped local subject identifier' flag and get roles from userStore.
+             */
+            if (isSPAlwaysSendMappedLocalSubjectId(clientId)) {
+                userRoles = getUserRoles(user);
+            } else {
+                // Handle not account associated federated users.
+                userRoles = getUserRolesForNotAssociatedFederatedUser(user);
             }
-            if (!isUserAuthorizedForScope(scope, userRoles, tenantId)) {
-                return false;
+        } else {
+            userRoles = getUserRoles(user);
+        }
+        if (ArrayUtils.isNotEmpty(userRoles)) {
+            for (String scope : requestedScopes) {
```

Comment:
Can't we use forEach() instead?

Reply:
forEach method works with Java lists, here we are looping object arrays. IMO, it's optimized. please correct me if I am wrong.